File: src/dotnet/Src/ButterflyNet.Parser/Strategies/Eol/ParagraphClosingStrategy.cs
using System;

namespace ButterflyNet.Parser.Strategies.Eol {
	public sealed class ParagraphClosingStrategy : IEolScopeClosingStrategy {
		public Type ScopeType { get { return ScopeTypeCache.Paragraph; } }

		public bool ShouldClose(ParseContext context) {
			var peek = context.Input.Peek();
			return context.Input.IsEof || peek == ButterflyStringReader.NoValue || peek == '\n';
		}
	}
}

File: src/js/src/strategies/eol/ParagraphClosingStrategy.js
function ParagraphClosingStrategy() {
	this.scopeType = ScopeTypeCache.paragraph;
	this.shouldClose = function(context) {
		var peek = context.input.peek();
		return context.input.isEof() || peek === ButterflyStringReader.EOF || peek === "\n";
	};
}